### PR TITLE
MAINT: migrate `special.nbdtrin` to boost 

### PR DESCRIPTION
**Pull request by JacobHass8**
Towards #21966. 

#### What does this implement/fix?
Migrates `nbdtrin` from cdflib to boost.

### CODE REVIEW

**Comment by JacobHass8** [comment]
@dschmitz89, I thought since I was implementing inverse functions in boost, I would try to migrate other functions from cdflib to boost. I based it on some of the previous PRs in #21966. Is there anything that I missed? There doesn't appear to be a lot of testing for this function. 

If this looks good, I can work on migrating other functions in cdflib to boost.

File: scipy/special/boost_special_functions.h [hunk 1]
```diff
@@ -1959,6 +1960,51 @@ nbinom_kurtosis_excess_double(double r, double p)
...
+
+    if (std::isnan(k) || std::isnan(y) || std::isnan(p)) {
+        return NAN;
+    }
+    if ((k <= 0) || (y < 0) || (y > 1) || (p < 0) || (p > 1)) {
+        sf_error("nbdtrik", SF_ERROR_DOMAIN, NULL);
+        return NAN;
+    }
+    try {
+        num_trials = boost::math::negative_binomial_distribution<Real>::find_minimum_number_of_trials(k, p, y);
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h]:
```suggestion
        num_trials = boost::math::negative_binomial_distribution<Real, SpecialPolicy()>::find_minimum_number_of_trials(k, p, y);
```

File: scipy/special/boost_special_functions.h [hunk 2]
```diff
@@ -1959,6 +1960,51 @@ nbinom_kurtosis_excess_double(double r, double p)
...
+        sf_error("nbdtrik", SF_ERROR_OVERFLOW, NULL);
+        n = INFINITY;
+    } catch (const std::underflow_error& e) {
+        sf_error("nbdtrik", SF_ERROR_UNDERFLOW, NULL);
+        n = 0;
+    } catch (...) {
+        sf_error("nbdtrik", SF_ERROR_OTHER, NULL);
+        n = NAN;
+    }
+    return n;
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h:1996]:
```suggestion
    if (n < 0) {
        n = NAN;
    }
    return n;
```

We have seen that boost can sometimes produce obscure results outside of the meant domain for edge cases, so I always include a check of the output now.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for this @JacobHass8 !
I was working on this as well actually but you beat me to it.

Two inline comments and two bigger ones. I will approve the CI run here once new tests are implemented.

*Tests*
This function is not tested in the `test_cdflib` file yet per https://github.com/scipy/scipy/blob/20460b6480e3cb3a06d6ab2241a8f636c8fb6e01/scipy/special/tests/test_cdflib.py#L10 so we need to add a hand full of spot checks using mpmath references. What I typically do is to calculate a reference for the CDF using mpmath and check that the inverse we implement recovers the input parameter.

The `test_basic` test file contains obscure ancient tests that were likely meant to just check that the function simply executes. `nbdtrin` is included there in https://github.com/scipy/scipy/blob/20460b6480e3cb3a06d6ab2241a8f636c8fb6e01/scipy/special/tests/test_basic.py#L702 . I expect this test to fail actually, it should simply be removed.

*Housekeeping*
We should already remove the unused old `nbdtrin` code from cdflib and the cython wrapper to avoid keeping bloat code around. You can check for example my [bdtrin PR](https://github.com/scipy/scipy/pull/24079) to find the relevant files.

*Bonus*
PRs for special functions often include an accuracy benchmark of the old versus the new implementation: see for example [here](https://github.com/scipy/scipy/pull/22699#issuecomment-2906689996). For me personally, this is not anymore a strict requirement for the cdflib -> boost migration as boost has proven to consistently outperform cdflib but other maintainers might ask for it.

File: scipy/special/boost_special_functions.h [hunk 3]
```diff
@@ -13,6 +13,7 @@
 
 #include "boost/math/distributions.hpp"
 #include <boost/math/distributions/inverse_gaussian.hpp>
+#include <boost/math/distributions/negative_binomial.hpp>
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h]: Is this include needed? We do not include all other distributions explicitly.

**Comment by JacobHass8** [comment]
> _Tests_ This function is not tested in the `test_cdflib` file yet per
> 
> https://github.com/scipy/scipy/blob/20460b6480e3cb3a06d6ab2241a8f636c8fb6e01/scipy/special/tests/test_cdflib.py#L10
> so we need to add a hand full of spot checks using mpmath references. What I typically do is to calculate a reference for the CDF using mpmath and check that the inverse we implement recovers the input parameter.

I've added some initial spot checks using mpmath and domain testing. I have a couple questions regarding edge cases though. What should happen when $y=0, 1$, $p=0,1$ or $k=\infty$? Currently: 

- `nbdtrin(k, 0, p) = inf`, 
- `nbdtrin(k,1,p) = 0`, 
- `nbdtrin(k, y, 0) = nan`, 
- `nbdtrin(k, y, 1) = nan` 
- `nbdtrin(inf, y, p) = nan`

 I'm not entirely sure what the correct behavior should be.

**Comment by dschmitz89** [comment]
> > _Tests_ This function is not tested in the `test_cdflib` file yet per
> > https://github.com/scipy/scipy/blob/20460b6480e3cb3a06d6ab2241a8f636c8fb6e01/scipy/special/tests/test_cdflib.py#L10
> > 
> > so we need to add a hand full of spot checks using mpmath references. What I typically do is to calculate a reference for the CDF using mpmath and check that the inverse we implement recovers the input parameter.
> 
> I've added some initial spot checks using mpmath and domain testing. I have a couple questions regarding edge cases though. What should happen when y = 0 , 1 , p = 0 , 1 or k = ∞ ? Currently:
> 
> * `nbdtrin(k, 0, p) = inf`,
> * `nbdtrin(k,1,p) = 0`,
> * `nbdtrin(k, y, 0) = nan`,
> * `nbdtrin(k, y, 1) = nan`
> * `nbdtrin(inf, y, p) = nan`
> 
> I'm not entirely sure what the correct behavior should be.

For the moment, when in doubt, stick with the current behaviour in main to maintain backwards compatability I would suggest. For me edge case handling is a battle for another day after #21966 is completed.

File: scipy/special/tests/test_cdflib.py [hunk 4]
```diff
@@ -718,6 +720,38 @@ def test_nctdtrit(self, df, nc, x, expected_cdf):
...
+            "k, y, p, n, rtol",
+            [(5, 0.3214569091796875, 0.25, 3, 1e-18),
+             (10, 3.820379007878081089302022753626e-70, 0.15, 100, 1e-15),
+             (100, 0.97557472292704298255653906603083, 0.15, 10, 1e-14),
+             (2, 3.5032532500000018230123192269914e-73, 0.001, 25, 1e-18),
+             (2, 0.99999932710299546878804333774717, 0.999, 15, 1e-10),
+             (500, 0.99999999986031784927135792193089, 0.1, 15, 1e-8),
+             (0, 1.0000000000000005551115123125784e-10, 0.1, 10, 1e-18)])
+    def test_inverse_n(self, k, y, p, n, rtol):
+        assert_allclose(sp.nbdtrin(k, y, p), n, rtol)
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py]: We should only use double precision numbers for test values, so could you reduce the number of decimal digits? 

Also, please include the code you used to calculate the reference values.

Comment (JacobHass8) [reply scipy/special/tests/test_cdflib.py]: From what I understand, double precision numbers get anywhere between 15-17 significant digits. I saw the `TestNoncentralChiSquaredFunctions` test class used 20 digits, so I just used 20 if that's okay. I also added the code I used to generate these values.

**Comment by JacobHass8** [comment]
> For the moment, when in doubt, stick with the current behaviour in main to maintain backwards compatability I would suggest. For me edge case handling is a battle for another day after #21966 is completed.

I'm not sure there is well defined behavior in main. Here are the values that I'm getting when running on the main branch.

- `nbdtrin(k, 0, p) = 1957.499999902125` (this value changes based on `k` and `p`, but sticks at this point for most values. I think the behavior here is not intended.), 
- `nbdtrin(k,1,p) = nan`, 
- `nbdtrin(k, y, 0) = 0`, 
- `nbdtrin(k, y, 1) = 1e100` 
- `nbdtrin(inf, y, p) = 1e100`

File: scipy/special/tests/test_cdflib.py (hunk 4 above)
Comment (dschmitz89) [reply scipy/special/tests/test_cdflib.py]: I guess you are referring to [these values](https://github.com/JacobHass8/scipy/blob/b90c4536f40a6db9a12a938d4613a343c397c9f2/scipy/special/tests/test_cdflib.py#L777). Those were computed with Wolfram Alpha a looong time ago with no attention to the number of digits. For most other tests, we stay within 17 digits.

You could simply do `float()` to convert your mpmath result to double precision (at least thats how I do it and no one complained so far).

**Comment by dschmitz89** [comment]
> > For the moment, when in doubt, stick with the current behaviour in main to maintain backwards compatability I would suggest. For me edge case handling is a battle for another day after #21966 is completed.
> 
> I'm not sure there is well defined behavior in main. Here are the values that I'm getting when running on the main branch.
> 
> * `nbdtrin(k, 0, p) = 1957.499999902125` (this value changes based on `k` and `p`, but sticks at this point for most values. I think the behavior here is not intended.),
> * `nbdtrin(k,1,p) = nan`,
> * `nbdtrin(k, y, 0) = 0`,
> * `nbdtrin(k, y, 1) = 1e100`
> * `nbdtrin(inf, y, p) = 1e100`

The `1e100` are a "feature" of cdflib representing the maximal value it can compute. Those two entries we could replace. Suggestions:

* `nbdtrin(inf, y, p) = inf` (n should be more than k, so inf is the only option?)
* `nbdtrin(k, y, 1) = inf` ? (if I understand correctly, the likelihood for the negative result is zero, so it is anyway impossible to get any meaningful result?)

About the first three, I do not have a strong opinion. Feel free to make an argument for either case. I do not think that it is super important anyway.

File: scipy/special/tests/test_cdflib.py (hunk 4 above)
Comment (JacobHass8) [reply scipy/special/tests/test_cdflib.py]: I just used float to convert to double precision! All the tests are still passing.

**Comment by JacobHass8** [comment]
>     * `nbdtrin(inf, y, p) = inf` (n should be more than k, so inf is the only option?)

I think that `n` is just the number of successes so it can be larger than `k`. If $k = \infty$, then $y$ should be 1 as $y$ is the cdf of the distribution. I don't think that we can extract any information in this case. 

>     * `nbdtrin(k, y, 1) = inf` ? (if I understand correctly, the likelihood for the negative result is zero, so it is anyway impossible to get any meaningful result?)

Yes, in this case, the CDF should always be 1 and we can't extract anything. I think the same goes for $p=0$ because then the CDF should always be 0. 

Now that I think about these edge cases, the CDF (`y`) will always be either 0 or 1. I don't think that we can extract any information about $n$ in these cases so I'd argue returning `nan` for _all_ of the edge cases above. I'm fine addressing this later though and using

 * `nbdtrin(k, 0, p) = nan` 
 * `nbdtrin(k,1,p) = nan`,
 * `nbdtrin(k, y, 0) = 0`,
 * `nbdtrin(k, y, 1) = inf`
 * `nbdtrin(inf, y, p) = inf`

to be backwards compatible. 

Note, for the binomial distribution, I found 
 * `bdtrin(k, 0, p) = inf` 
 * `bdtrin(k,1,p) = k`,
 * `bdtrin(k, y, 0) = nan`,
 * `bdtrin(k, y, 1) = nan`
 * `bdtrin(inf, y, p) = nan`

**Comment by dschmitz89** [comment]
> >     * `nbdtrin(inf, y, p) = inf` (n should be more than k, so inf is the only option?)
> 
> I think that `n` is just the number of successes so it can be larger than `k`. If $k = \infty$, then $y$ should be 1 as $y$ is the cdf of the distribution. I don't think that we can extract any information in this case. 
> 
> >     * `nbdtrin(k, y, 1) = inf` ? (if I understand correctly, the likelihood for the negative result is zero, so it is anyway impossible to get any meaningful result?)
> 
> Yes, in this case, the CDF should always be 1 and we can't extract anything. I think the same goes for $p=0$ because then the CDF should always be 0. 
> 
> Now that I think about these edge cases, the CDF (`y`) will always be either 0 or 1. I don't think that we can extract any information about $n$ in these cases so I'd argue returning `nan` for _all_ of the edge cases above. I'm fine addressing this later though and using
> 
>  * `nbdtrin(k, 0, p) = nan` 
>  * `nbdtrin(k,1,p) = nan`,
>  * `nbdtrin(k, y, 0) = 0`,
>  * `nbdtrin(k, y, 1) = inf`
>  * `nbdtrin(inf, y, p) = inf`
> 
> to be backwards compatible. 
> 
> Note, for the binomial distribution, I found 
>  * `bdtrin(k, 0, p) = inf` 
>  * `bdtrin(k,1,p) = k`,
>  * `bdtrin(k, y, 0) = nan`,
>  * `bdtrin(k, y, 1) = nan`
>  * `bdtrin(inf, y, p) = nan`
> 
> 

Going with `nan` for all of them is fine for me. I recently migrated `bdtrin` to boost and did not include additional edge case handling. I do not find it super important for such niche functions.

**Comment by JacobHass8** [comment]
> Going with `nan` for all of them is fine for me. I recently migrated `bdtrin` to boost and did not include additional edge case handling. I do not find it super important for such niche functions.

I pushed a commit which ensures all the edge cases return nan. Are there any other issues that you see?

File: scipy/special/tests/test_cdflib.py [hunk 5]
```diff
@@ -718,6 +717,53 @@ def test_nctdtrit(self, df, nc, x, expected_cdf):
...
+
+    @pytest.mark.parametrize(
+            "k, y, p, n, rtol",
+            [(5, 0.3214569091796875, 0.25, 3, 1e-18),
+             (10, 3.820379007878081e-70, 0.15, 100, 1e-15),
+             (100, 0.975574722927043, 0.15, 10, 1e-14),
+             (2, 3.503253250000002e-73, 0.001, 25, 1e-18),
+             (2, 0.9999993271029954, 0.999, 15, 1e-10),
+             (500, 0.9999999998603178, 0.1, 15, 1e-8),
+             (0, 1.0000000000000006e-10, 0.1, 10, 1e-18)])
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py]:
```suggestion
             (0, 1.0000000000000006e-10, 0.1, 10, 5e-16)])
```

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
My suggestion should fix the last test failures we see on Windows in CI.

Lastly, we should also remove the formula and the book reference from the docs. Interested users should check out the boost documentation for details instead.

**Comment by j-bowhay** [comment]
> Lastly, we should also remove the formula and the book reference from the docs. Interested users should check out the boost documentation for details instead.

Is there a strong reason to remove these references? They look potentially handy to me:)

**Comment by dschmitz89** [comment]
> > Lastly, we should also remove the formula and the book reference from the docs. Interested users should check out the boost documentation for details instead.
> 
> Is there a strong reason to remove these references? They look potentially handy to me:)

I don't really have strong feelings about it. My main point is that our docs will likely get out of sync with upstream boost. But not a blocker.

**Review by dschmitz89 — APPROVED** [review]
Allright, will give it a few days before merging.

Thanks again @JacobHass8 .

PS: regarding #21966 I am currently looking into `nrdtrimn` and `nrdtrisd`. After those and `nbdtrik`, we are only missing the ones we need to add to boost. If you are interested to help out more with this issue: as you already made good progress with one boost PR, would you consider opening more PRs in boost?

**Comment by JacobHass8** [comment]
> If you are interested to help out more with this issue: as you already made good progress with one boost PR, would you consider opening more PRs in boost?

Yes, I can implement the remainder of the needed functions in boost! I was going to use the first PR as a template before opening PRs for the remaining functions.

File: scipy/special/boost_special_functions.h [hunk 6]
```diff
@@ -1959,6 +1959,55 @@ nbinom_kurtosis_excess_double(double r, double p)
...
+{
+    Real n;
+    Real num_trials;
+
+    if (std::isnan(k) || std::isnan(y) || std::isnan(p)) {
+        return NAN;
+    }
+    // The edge cases of p=0,1 is handled in boost 
+    if ((k < 0) || (y <= 0) || (y >= 1) || (p < 0) || (p > 1)) {
+        sf_error("nbdtrik", SF_ERROR_DOMAIN, NULL);
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h]:
```suggestion
        sf_error("nbdtrin", SF_ERROR_DOMAIN, NULL);
```

File: scipy/special/boost_special_functions.h [hunk 7]
```diff
@@ -1959,6 +1959,55 @@ nbinom_kurtosis_excess_double(double r, double p)
...
+        return NAN;
+    }
+    try {
+        num_trials = boost::math::negative_binomial_distribution<Real, SpecialPolicy>::find_minimum_number_of_trials(k, p, y);
+        n = num_trials - k;
+    } catch (const std::domain_error& e) {
+        sf_error("nbdtrik", SF_ERROR_DOMAIN, NULL);
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h]:
```suggestion
        sf_error("nbdtrin", SF_ERROR_DOMAIN, NULL);
```

File: scipy/special/boost_special_functions.h [hunk 8]
```diff
@@ -1959,6 +1959,55 @@ nbinom_kurtosis_excess_double(double r, double p)
...
+        n = NAN;
+    } catch (const std::overflow_error& e) {
+        sf_error("nbdtrik", SF_ERROR_OVERFLOW, NULL);
+        n = INFINITY;
+    } catch (const std::underflow_error& e) {
+        sf_error("nbdtrik", SF_ERROR_UNDERFLOW, NULL);
+        n = 0;
+    } catch (...) {
+        sf_error("nbdtrik", SF_ERROR_OTHER, NULL);
```
Comment (dschmitz89) [inline scipy/special/boost_special_functions.h]:
```suggestion
        sf_error("nbdtrin", SF_ERROR_OVERFLOW, NULL);
        n = INFINITY;
    } catch (const std::underflow_error& e) {
        sf_error("nbdtrin", SF_ERROR_UNDERFLOW, NULL);
        n = 0;
    } catch (...) {
        sf_error("nbdtrin", SF_ERROR_OTHER, NULL);
```

File: scipy/special/tests/test_cdflib.py [hunk 9]
```diff
@@ -718,6 +717,53 @@ def test_nctdtrit(self, df, nc, x, expected_cdf):
...
+
+    @pytest.mark.parametrize(
+            "k, y, p, n, rtol",
+            [(5, 0.3214569091796875, 0.25, 3, 1e-18),
+             (10, 3.820379007878081e-70, 0.15, 100, 1e-15),
+             (100, 0.975574722927043, 0.15, 10, 1e-14),
+             (2, 3.503253250000002e-73, 0.001, 25, 1e-18),
+             (2, 0.9999993271029954, 0.999, 15, 1e-10),
+             (500, 0.9999999998603178, 0.1, 15, 1e-8),
+             (0, 1.0000000000000006e-10, 0.1, 10, 5e-16)])
```
Comment (dschmitz89) [inline scipy/special/tests/test_cdflib.py]:
```suggestion
            [(5, 0.3214569091796875, 0.25, 3, 5e-16),
             (10, 3.820379007878081e-70, 0.15, 100, 1e-15),
             (100, 0.975574722927043, 0.15, 10, 1e-14),
             (2, 3.503253250000002e-73, 0.001, 25, 5e-16),
             (2, 0.9999993271029954, 0.999, 15, 1e-10),
             (500, 0.9999999998603178, 0.1, 15, 1e-8),
             (0, 1.0000000000000006e-10, 0.1, 10, 5e-16)])
```

**Comment by dschmitz89** [comment]
No need to waste GPU CI cycles here imo, so merging.

Thanks @JacobHass8 !
